src/operator/rnn-inl.h

```diff
@@ -1360,15 +1360,16 @@ class RNNOp {
       // Create Dropout descriptors
       if (param_.p > 0) {
          ctx.requested[rnn_enum::kCuDNNDropoutDescSpace].get_cudnn_dropout_desc
-            (&dropout_desc_, s, 1.0f - param_.p, seed_);
+            (&dropout_desc_, s, 1.0f - param_.p);
       }
       // Only update the probability by passing in a null dropout_states ptr
       DType* dropout_states = NULL;
       size_t dropout_bytes = 0;
+      // use dummy seed as state is null
       CUDNN_CALL(cudnnSetDropoutDescriptor(dropout_desc_, s->dnn_handle_,
                                            param_.p,  // discard probability
                                            dropout_states, dropout_bytes,
-                                           seed_));
+                                           0));
```

Comment:
In the minimum, the comment should be updated.  The way these calls work is:
cudnnSetDropoutDescriptor(..., dropout_states==NULL,...)    // Set dropout probability **and seed**, leave states alone.
cudnnSetDropoutDescriptor(..., dropout_states!=NULL,...)    // Set dropout probability and seed, init states based on these values.

cudnnRestoreDropoutDescriptor()    // Set dropout probability, seed and states ptr from provided args.

Reply:
Ok, and I found another issue, the seed will be fixed during resource initialization. So once if dropout layer is created, event if we don't fix  seed, each dropout result will be the same.

If we want each forward to have a different result, and the random result must respect mxnet random seed. The only solution is to get mxnet seed during each forward. Is that right?

Reply:
new test case
```
@with_seed()
def test_dropout_with_seed():
    info = np.iinfo(np.int32)
    seed = np.random.randint(info.min, info.max)
    _test_dropout(mx.cpu(), seed)
    _test_dropout(mx.gpu(), seed)
    _test_dropout(mx.cpu())
    _test_dropout(mx.gpu())

def _test_dropout(ctx, seed=None):
    data = mx.nd.ones((100, 100), ctx=ctx)
    dropout = mx.gluon.nn.Dropout(0.5)

    if seed:
        mx.random.seed(seed)
    with mx.autograd.record():
        result1 = dropout(data)

    if seed:
        mx.random.seed(seed)
    with mx.autograd.record():
        result2 = dropout(data)
    if seed:
        # dropout on gpu should return same result with fixed seed
        assert_almost_equal(result1.asnumpy(), result2.asnumpy())
    else:
        # seed not fixed, result should be different
        with assert_raises(AssertionError):
            assert_almost_equal(result1.asnumpy(), result2.asnumpy())
```

Reply:
Dick is looking into it, but I don't think sentence `event if we don't fix seed, each dropout result will be the same` is true. What is true though, is the fact that the dropout as is implemented now would not respond to the change of seed on the MXNet side after the initial creation of the op. Seeding each forward would destroy performance, so how about a solution like this - if you know that you will cache the value of seed (like in the dropout descriptor resource case), every time when you get the descriptor it should internally ask the random resource "did a user set a new seed?" (which could be implemented as a set in the random resource that would keep track of who asked about this already and resetting that set when user calls to `mxnet.random.seed`). If the answer is "no", then no reseeding is required, but if the answer is yes, the dropout descriptor should be reseeded with a new value.

Reply:
echoing @ptrendx 's comment, we should implement SetSeed() for the cudnn resource in  https://github.com/apache/incubator-mxnet/blob/master/src/resource.cc#L174-L201